### ISSUE TITLE
How do we implement exception handling around Command.Invoke?

### ISSUE THREAD

**Issue by SuperJMN** [dotnet/command-line-api#796]
I have created a simple .NET Core console application with this line to execute the `Command`
```
try
{   
    rootCommand.Invoke(args)
}
catch 
{
    Console.Error.Write("Something bad happened");
}
```

But when the command handler associated to the command throws an exception, the catch block isn't executed. Instead, the Invoke method returns a non-zero integer.

So, what's the correct method to handle exceptions?

**Comment by jonsequitur** [comment]
When you invoke a command directly like this, it gets wrapped with the default middlewares including `UseExceptionHandler`. This is intended to handle exceptions very generally but if you want to handle more specific exceptions, you should do so inside your command handler, which gets called after the middleware.

**Comment by AndrewSav** [comment]
Yes, this is something that I tripped on too. In my opinion printing the stack trace to the end user by default is not reasonable. Asking to create a separate exception handler for each command does not seem to conform to DRY either. I do not expect any library to catch and swallow my exception, I prefer to process them myself. A helper option like `UseExceptionHandler` is nice, but that should not be the default. 

The workaround seems to be:

```
            var _ = new CommandLineBuilder(rootCommand)
                   .UseVersionOption()
                   .UseHelp()
                   .UseEnvironmentVariableDirective()
                   .UseParseDirective()
                   .UseDebugDirective()
                   .UseSuggestDirective()
                   .RegisterWithDotnetSuggest()
                   .UseTypoCorrections()
                   .UseParseErrorReporting()
                   .CancelOnProcessTermination()
                   .Build();

                   rootCommand.Invoke(args);
```

Note, that the result of the `CommandLineBuilder` is discarded, it is stored in `ImplicitParser` property on the command object itself. Not very intuitive, but I guess could be documented?

**Comment by ohadschn** [comment]
Agree with @AndrewSav - IMHO `UseExceptionHandler` should not be on by default (or at the very least have a simple flag to turn it off), I prefer to process my exceptions too. As one example, I like to let fatal exceptions (like out of memory) pass through and crash the process.

**Comment by jonsequitur** [comment]
@AndrewSav Typically if you're wiring up the `CommandLineBuilder` our assumption is you'll call `Invoke` on the parser returned by its `Build` method rather than using the `Command.Invoke` extension method.

**Comment by jonsequitur** [comment]
As a middleware, `UseExceptionHandler` applies to all commands, so it satisfies DRY. You can customize it by passing your own delegate if you want to change the default behaviors, which are:

1. Sets the result code to `1`.
2. Prints a stack trace to `Console.Error`.

We're open to suggestions that change these defaults. So far, printing the stack trace is the most useful generalizable approach we've seen suggested. Here's the reasoning. Your program crashed in some way your code didn't anticipate. An uninformative general message means that in order for you to investigate, either 1) your user would have to debug the program or 2) you would need to implement telemetry. The stack trace is trivial for your user to copy and paste into an issue, or search for online in order to find solutions. If the exception handler isn't in place, the just-in-time debugger will kick in for users who have Visual Studio installed. This is also more likely to happen to your users than to you, and they probably won't debug your code.

Do you have a preferred behavior in mind?

**Comment by AndrewSav** [comment]
> As a middleware, UseExceptionHandler applies to all commands, so it satisfies DRY.

Sorry, my comment regarding DRY was specifically in response to suggestion of catching exceptions in command handlers. This was not with regards to the middleware.

> We're open to suggestions that change these defaults. So far, printing the stack trace is the most useful generalizable approach we've seen suggested.

The suggestion was to not enable this middleware by default. Once one has decided that they want to use the middleware, the in-middleware defaults are indeed reasonable - I did not object to these.

**Comment by jonsequitur** [comment]
Not enabling it by default means the default behavior on an unhandled exception is a process crash, a stack trace, and possible triggering of the just-in-time debugger. It's also not a great experience, and perhaps slightly less useful.

**Comment by AndrewSav** [comment]
I disagree. CommandLine is a library, in my opinion a library should not catch exceptions it did not produce. If you change the handler to only catch exception that may have originate from CommandLine library itself, that would make certain sense. Swallowing unrelated exceptions should not happen.

There are also always "generic" catastrophic failure exceptions such as out of memory condition, those in my opinion should not be swallowed either even if they were caused by an allocation within the library.

I'm not alone. [Phil Haack](https://haacked.com/) wrote in a comment to [this Hanselman's page](https://www.hanselman.com/blog/GoodExceptionManagementRulesOfThumb.aspx):

> I'd probably mention that there is a difference between how you handle exceptions when writing a library vs writing an application. For example, a library should generally just propagate exceptions unless it's absolutely sure it can handle it in the best way possible.

I think @ohadschn and @SuperJMN  above would also agree with this sentiment.

**Comment by jonsequitur** [comment]
It's often said that the difference between a library and a framework is that you call a library whereas a framework calls you. By that definition anyway, some parts of System.CommandLine are library (e.g. the parser) and some are framework (e.g. handler invocation). It's influenced by ASP.NET which, like many frameworks, has a top-level catch-all exception handler. The purpose of this is to handle exceptions that would affect the user experience in the cases when they were not handled by the developer. 

It's also the case that the default exception handler behavior was based on user feedback that a crash is not a good default user experience. Knowing we can't design something that's perfect for every use case, we made the API composable. You can opt out of the default exception handling behavior by not calling `UseDefaults` or `UseExceptionHandler`.

If you have suggestions to improve the default experience, we want to hear them along with your opinion on why it's an improvement. We do have a bias toward user experience, and I think that's the piece I'm missing. How does removing the default exception handler improve user experience?

It also might be worth opening a separate issue, as @SuperJMN was asking a question and not, as far as I can tell, suggesting a design change.

**Comment by AndrewSav** [comment]
@jonsequitur, thank you for engaging in this discussion, I really appreciate your time. I feel that we have different ideological priorities, which will not be aligned in this thread. It appears that you have given a careful consideration to the issues that has not changed the outcome.

Command Line parsing has always been a bit of a problem in the .Net world. There has been many attempts to make it easier by different library, and none took on enough to become the mainstream. I was hopeful when CommandLine first emerged as I saw this as an opportunity to finally get it right. I wish it to succeed and make life of countless developers easier.

>If you have suggestions to improve the default experience, we want to hear them along with your opinion on why it's an improvement. We do have a bias toward user experience, and I think that's the piece I'm missing. How does removing the default exception handler improve user experience?

When we talk about CommandLine parsing library, what it does is helping a _developer_ to spend less time on a boilerplate code. Writing a good command line parsing library is surprisingly difficult, I would identify the main challenge that it's a futile attempt to make it cater for all possible command line syntax's, yet it should not make a developer feel shoehorned into a particular one that does not match his need. I think CommandLine strikes a good balance here.

In my mind the focus is on the _developer experience_ not on the end-user experience. We have developers to care for end-user experience, so let's leave it with them. Let's empower them to provide that experience quickly and easily. We, as the library designers will focus on developers experience, so that developers could easily deliver top notch user experience.

Let's say I as a developer have a code that throws and unhanded exception under some circumstances. Is this a good user experience? Perhaps not. So to help that I catch this exception and handle it somewhere. Now I realised that my quick and dirty home-grown command line parser is inadequate, so I quickly plug-in CommandLine. This is just a library, so I do not expect it to interfere with exception handling. And yet it does!

My _developer experience_ is not very good, It is not obvious at all what's happening or why, and it's not obvious how to work around that.

User experience has changed to _worse_ an exception used to be nicely handled, and now it is dumped on screen and scary the user with technical stuff they do not understand.

In my mind it's pretty clear that this should not be the default. Even in asp.net framework I was not happy with that default, but for CommandLine, in my opinion it's far out.

But I do understand that you don't see it this way, and it does not appear that I can persuade you. So I'll leave it at this. Once again, thank you for contributing to this discussion, it's much appreciated.

**Comment by ohadschn** [comment]
Agree with @AndrewSav and IMHO `System.CommandLine` is not a framework (I don't think having a handler makes it one).

At any rate, I think we can all live with the current default, as long as it's easy to opt out of it (and keep all the other good default stuff). Specifically, it has to be easier than:
```cs
var _ = new CommandLineBuilder(rootCommand)
       .UseVersionOption()
       .UseHelp()
       .UseEnvironmentVariableDirective()
       .UseParseDirective()
       .UseDebugDirective()
       .UseSuggestDirective()
       .RegisterWithDotnetSuggest()
       .UseTypoCorrections()
       .UseParseErrorReporting()
       .CancelOnProcessTermination()
       .Build();
```
Not just easier, but immune to future default changes (e.g. added directives and whatnot). Something that works with the most basic sample program like:
```cs
rootCommand.Handler = CommandHandler.Create<int>(handleExceptions: false, (intOption) =>
    {
        Console.WriteLine($"The value for --int-option is: {intOption}");
    });
```

**Comment by jnm2** [comment]
I would really like to be able to handle exceptions myself without 1) repeating the handling in the root of each invoked command or 2) writing more than one line of code to opt out of the default behavior.

**Comment by jnm2** [comment]
This is what I'm doing to minimize repetition. I still have to repeat the `.WithExceptionHandler(ExceptionHandler)` call once per subcommand, but it's better than hardcoding what the CommandLineBuilder defaults were at this point in time:

```cs
/// <summary>
/// Workaround for <see href="https://github.com/dotnet/command-line-api/issues/796"/>.
/// </summary>
public static ICommandHandler WithExceptionHandler(this ICommandHandler commandHandler, Func<Exception, int> exceptionHandler)
{
    return new ExceptionHandlingCommandHandler(commandHandler, exceptionHandler);
}

private sealed class ExceptionHandlingCommandHandler : ICommandHandler
{
    private readonly ICommandHandler commandHandler;
    private readonly Func<Exception, int> exceptionHandler;

    public ExceptionHandlingCommandHandler(ICommandHandler commandHandler, Func<Exception, int> exceptionHandler)
    {
        this.commandHandler = commandHandler;
        this.exceptionHandler = exceptionHandler;
    }

    public async Task<int> InvokeAsync(InvocationContext context)
    {
        try
        {
            return await commandHandler.InvokeAsync(context);
        }
        catch (Exception ex)
        {
            return exceptionHandler.Invoke(ex);
        }
    }
}
```

**Comment by jonsequitur** [comment]
Any suggestions on an API for opting out of this?

A related problem that needs to be solved is how extension methods like `Command.Invoke` should behave, given that under the hood they use `UseDefaults` so that people don't need to use the `CommandLineBuilder`.

@jnm2, I'm curious about why you implemented a custom `ICommandHandler` instead of a middleware.

**Comment by jnm2** [comment]
@jonsequitur I haven't figured out middleware yet. It's odd that you wouldn't be able to add middleware to a root command using an API exposed on the root command instance.

**Comment by jnm2** [comment]
Can we have a Configuration property on the root command and set `UseDefaultExceptionHandler = false`?

**Comment by jonsequitur** [comment]
The `Symbol` types don't know anything about the details of invocation (other than having a pointer to a handler). We wouldn't want to duplicate invocation functionality into `Command`.

The invocation, including the current exception handling, is all implemented in middleware, which gets called before and after the command handler is invoked.

The current default exception handling is implemented like this:

```csharp
        public static CommandLineBuilder UseExceptionHandler(
            this CommandLineBuilder builder,
            Action<Exception, InvocationContext>? onException = null)
        {
            builder.AddMiddleware(async (context, next) =>
            {
                try
                {
                    await next(context);
                }
                catch (Exception exception)
                {
                    (onException ?? Default)(exception, context);
                }
            }, MiddlewareOrderInternal.ExceptionHandler);

            return builder;

            void Default(Exception exception, InvocationContext context)
            {
                context.Console.ResetTerminalForegroundColor();
                context.Console.SetTerminalForegroundRed();

                context.Console.Error.Write("Unhandled exception: ");
                context.Console.Error.WriteLine(exception.ToString());

                context.Console.ResetTerminalForegroundColor();

                context.ResultCode = 1;
            }
        }
```

**Comment by jonsequitur** [comment]
> It's odd that you wouldn't be able to add middleware to a root command using an API exposed on the root command instance.

There's a single middleware pipeline per `CommandLineConfiguration`, not per command. The middleware is intended to implement cross-cutting behaviors.

**Comment by jnm2** [comment]
I think I figured out how to use middleware. Is this right?

```cs
return await new CommandLineBuilder(rootCommand)
    .UseDefaults()
    .UseExceptionHandler((exception, context) => context.ResultCode = ExceptionHandler(exception))
    .Command.InvokeAsync(args);
```

I still would like to use `try`...`catch` rather than passing a delegate to an exception handler.

> The Symbol types don't know anything about the details of invocation (other than having a pointer to a handler). We wouldn't want to duplicate invocation functionality into Command.

`Command` owns the state because `CommandLineBuilder` sets `Command.ImplicitParser`. I'd rather that the state is exposed transparently wherever it actually lives. If the thing determining how exception handling works is living inside a property inside the root command, wouldn't it make sense to let you access that and configure it?

> There's a single middleware pipeline per CommandLineConfiguration, not per Command. The middleware is intended to implement cross-cutting behaviors.

I think I understand this. There's only one root command though, so this could be exposed through RootCommand.

**Comment by jnm2** [comment]
If it's really imperative that the root command can't have configuration attached to it, what if I could do something like ASP.NET's configuration APIs?

```cs
return await rootCommand
    .Configure(config => config.UseDefaultExceptionHandler = false)
    .InvokeAsync(args);
```

**Comment by jonsequitur** [comment]
> Command owns the state because CommandLineBuilder sets Command.ImplicitParser.

This is an implementation detail which is subject to change. This is effectively just a cache.

> There's only one root command though, so this could be exposed through RootCommand. 

The `RootCommand` isn't involved in invocation of subcommands. It's probably clearer to think of them as orthogonal, independent routes (to use an MVC metaphor.) Just like in ASP.NET, our middleware is intentionally cross-cutting, and no special importance is given to the "root" command other than that its name is derived from the executable's name by default, i.e. it's `argv[0]`. If you're using the parser in a REPL, for example, you could omit `RootCommand` from your parser configuration.

**Comment by jonsequitur** [comment]
```csharp
return await rootCommand
    .Configure(config => config.UseDefaultExceptionHandler = false) // Returns some new type or a Parser instance maybe
    .InvokeAsync(args);
```

I think in isolation this API is pretty intuitive. How would it work when someone also does this?

```csharp
new CommandLineBuilder(rootCommand)
    .UseExceptionHandler()
    .Build();
```

**Comment by jnm2** [comment]
Wouldn't it work the same way as the builder already works? Last person to call Build() wins?

**Comment by jnm2** [comment]
`Command.InvokeAsync` already effectively performs a `new CommandLineBuilder(command)...Build()` behind the scenes (even if that's not literally what it does). I just want a way to pass parameters to that process such as `UseDefaultExceptionHandler = false`.

**Comment by jnm2** [comment]
I also don't honestly see a conceptual difference between:

```cs
var rootCommand = new RootCommand { subcommand1, subcommand2, ... };
rootCommand.Configuration.CrossCuttingStuff(...);
return await rootCommand.InvokeAsync(args);
```


```cs
var rootCommand = new RootCommand { subcommand1, subcommand2, ... };

new CommandLineBuilder(rootCommand)
    .CrossCuttingStuff(...);
    .Build();

return await rootCommand.InvokeAsync(args);
```

Except that the second one seems a lot less obvious to me. I'd even expect `rootCommand.InvokeAsync` to not be affected by CommandLineBuilder, and I'd expect to have to call `.Build().InvokeAsync(...)` in order to make use of the configuration I just did.

**Comment by jnm2** [comment]
I apologize for the quantity here, but I realized that I don't want the ability to use `try`...`catch` rather than passing a handler delegate. The reason is that the handler delegate has access to the IConsole and a `try`...`catch` around InvokeAsync does not.

**Comment by jonsequitur** [comment]
> Knowing what I know now, I also don't honestly see a conceptual difference between:

I'm not sure there is a meaningful one, but when using `CommandLineBuilder`, the guidance is to do this, rather than call `rootCommand.Invoke`:

```csharp
var rootCommand = new RootCommand { subcommand1, subcommand2, ... };

var parser = new CommandLineBuilder(rootCommand)
    .CrossCuttingStuff(...);
    .Build();

return await parser.InvokeAsync(args);
```

I think ideally, `CommandLineBuilder` should not mutate or store state on the `RootCommand` in any way. It might be a bug that `ImplicitParser` is set by calling the public `Build` method.

Currently, a `RootCommand` doesn't have a `1:1` relationship to a parser, middleware, binding rules, etc. and in fact doesn't know much about them. (Yes, it does have that `Handler` method, but it's pretty opaque, and `ImplicitParser` is internal and might be redesigned.) Adding `RootCommand.Configuration` would add a circular dependency between these layers. Leaving the design decoupled this way is intentional, so that the parser has the potential to ship as a separate component from the invocation and model binding layers, and other approaches to invocation and binding can be added later without having to overhaul the parser. 

Some other details to consider:

* You can call, for example, `new Command("x").Invoke("some args")` or `new Option<string>("-x").Parse()`. There's no `RootCommand` in the picture.

* Not all usages of the library use `RootCommand` (directly, though to some extent they sometimes have to work around it, but we'd like to make these scenarios easier).

**Comment by ohadschn** [comment]
> Any suggestions on an API for opting out of this?

The API could be as simple as `UseDefaults(bool handleExceptions)` and/or `UseDefaults(Action<Exception, InvocationContext> exceptionHandler)`. 

> A related problem that needs to be solved is how extension methods like `Command.Invoke` should behave, given that under the hood they use `UseDefaults` so that people don't need to use the `CommandLineBuilder`.

The same (optional) parameters could be added to the extensions methods and then passed on to their internal `UseDefaults` call, e.g. `Invoke(string[] args, bool handleExceptions)`. 

Another option, while we're taking pages off the ASP.NET Core book, could be something like `new CommandLineBuilder(rootCommand).UseDefaults().ReplaceExceptionHandler(Action<Exception, InvocationContext> exceptionHandler)`.

Yet another option would be some static default handler to be used by `UseDefaults`, e.g. `CommandLineBuilder.DefaultExceptionHandler=...`.

Finally, I think we may have lost sight a little bit of the original desire here - namely easily opt out or replace the default exception handling. It doesn't have to be perfect and cover all possible use cases, just allow those who want to an easy and reliable way to do it. For example even if I have to create my own `CommandLineBuilder` (that is I can't use the convenience `Command` extension methods discussed above), that's OK as I have at least one simple and reliable solution.

**Comment by AndrewSav** [comment]
So the solution I ended up with, in additions to the `new CommandLineBuilder(rootCommand).Use` multi-line monstrosity is this:

```
    static class CommandHandlerWrapper
    {
        public static async Task<int> ExecuteCommandHandler<T>(T options, Func<T, Task> handler)
        {
            try
            {
                await handler(options);
                return (int)ExitCode.Success;
            }
            catch (MyAppSpecificException ex)
            {
                return (int)ex.ExitCode;
            }
        }
    }
```

And then call it as:

```
command.Handler = CommandHandler.Create<SpecificOptions>(async specificOptions => await 
  CommandHandlerWrapper.ExecuteCommandHandler(specificOptions, CommandClass.Execute));
```
This avoids the try/catch repetition but introduces the unneeded wrapper,which we can do away with if we implement one of the suggestions above in the command-line-api.

Note that this approach shown is not generic, it requires a certain signature on the likes of `CommandClass.Execute` and it assumes a single `options` argument. It also assumes that if the handler does not throw it succeeds (which is correct in my case, but may be incorrect in general).

**Comment by pwretmo** [comment]
With the current behavior, what is the best way to dispose of resources created before invoking the root command? And in my case, flush and close Serilog to make sure unhandled exceptions are logged?

It seems to me that System.CommandLine is taking on responsibilities that is not related to executing commands, which makes the library / framework harder to use.

**Comment by SaintGimp** [comment]
This is an old issue but it's still open so I'll add my comment: I agree that the current API design is unexpected enough that I felt the need to write a big comment in my own code to explain what's going on:

```
// Goofy System.CommandLine API going on here.  We need to create and configure
// a builder in order to override its default exception handler and provide our own,
// except that we don't actually use the builder object or its return
// value.  Instead it modifies the root command and then we invoke that.
// https://github.com/dotnet/command-line-api/issues/796
var _ = new CommandLineBuilder(rootCommand)
    .UseDefaults()
    .UseExceptionHandler((e, context) =>
    {
        [custom exception handling here]
    }, 1)
    .Build();

await rootCommand.InvokeAsync(args);
```

The fact that there's an exception handler inserted by default was frustrating until I figured it out, but better System.CommandLine docs could help avoid that problem in the future.  The big API flaw that I'd like to see fixed before an official 2.0 release is that you have to use the CommandLineBuilder to change or remove the exception handler but you don't actually *use* the builder.  It just kind of sits there flapping around.  I'd like to see either a .Configure() chain on the RootCommand object instead of the builder, or to be able to run .InvokeAsync() on what CommandLineBuilder.Build() returns.  Either one would be self-describing.  What we've got now is a head-scratcher.

**Comment by AndrewSav** [comment]
> The big API flaw that I'd like to see fixed before an official 2.0 release is that you have to use the CommandLineBuilder to change or remove the exception handler but you don't actually use the builder. It just kind of sits there flapping around. 

Oh yeah, I found that funny too. Not major, but a code smell. I agree with all you just said. For an opportunity to create a brand new API those choices seemed strange. **jonsequitur** tried to [justify them above](https://github.com/dotnet/command-line-api/issues/796#issuecomment-672285614) but I'm not convinced. Having said that I have not followed what changed in the library in this respect in the last year, but judging by your comment not much.

**Comment by jonsequitur** [comment]
@SaintGimp The `_` surprised me. The more common usage of the `CommandLineBuilder` is to invoke through the `Parser` returned from `Build`: 

```cs
var parser = new CommandLineBuilder(rootCommand)
    .UseDefaults()
    .UseExceptionHandler((e, context) =>
    {
        [custom exception handling here]
    }, 1)
    .Build();

await parser.InvokeAsync(args);
```

The fact that your version works was unintentional.

Agreed, @AndrewSav, the fact that the `RootCommand` is mutated by the builder is a smell. This resulted from the `Command.Invoke`/`Command.InvokeAsync` convenience methods using a builder internally to set up the default behaviors. We realized every call was redoing that work, so we cached the `Parser`. 

Suggestions are welcome on what should be the correct behavior here, or how better to communicate these intentions through the API.

**Comment by SaintGimp** [comment]
@jonsequitur Aha, I see what's leading people (well, me at least) down the wrong path and into the ditch.  The Parse class doesn't have an InvokeAsync method.  That's actually an extension method and you have to have `using System.CommandLine.Parsing;` in order to get it.  If you aren't already using the System.CommandLine.Parsing namespace, then you'll think that parser.InvokeAsync() doesn't exist, but RootCommand.InvokeAsync() does, and you'll discover that the code I wrote above does actually work (though it looks goofy), and you'll end up thinking that's the only way to do things.

**Comment by jonsequitur** [comment]
It might make sense to move those extension methods into the `System.CommandLine` namespace so they'll be more discoverable.

**Comment by qui8t** [comment]
Using version `2.0.0-beta4.22272.1`, `.UseDefaults()` will cause the delegate of `.UseExceptionHandler` not invoked. 
The delegate of `UseExceptionHandler` is invoked only after `.UseDefaults()` is removed.

**Comment by qui8t** [comment]
Fingers crossed for the improved design. 

I can summarize my code but not sure if the errors would be reproducible since I'm not sure which parts of the code are the culprit (if I knew, I would have fixed them :)).  In general: 

```csharp
var rootCmd = new RootCommand();
rootCmd.AddOption(anOption);
rootCmd.SetHandler((options) => { Handler(options);},
    new OptionsBinder(anOption));

var parser = new CommandLineBuilder(rootCmd)
    .UseExceptionHandler((e, context) =>
    {
        Logger.LogException(e.Message);
    }, errorExitCode: 1)
    .UseVersionOption().UseEnvironmentVariableDirective().UseParseDirective().UseSuggestDirective() .RegisterWithDotnetSuggest().UseTypoCorrections().UseParseErrorReporting().CancelOnProcessTermination().UseTypoCorrections().UseHelp()
    .Build();

Environment.ExitCode = parser.Invoke(args);
```

Where: 

```csharp
class OptionsBinder : BinderBase<Options>
{
    private readonly Option<int> _anOO;
    public OptionsBinder(Option<int> anOptionOption)
    {
        _anOO = anOptionOption;
    }

    protected override Options GetBoundValue(BindingContext context)
    {
        var hasValue = context.ParseResult.FindResultFor(_anOO) != null;
        if (!hasValue) new Options();

        var value = context.ParseResult.GetValueForOption(_anOO);
        if (value is null)
            return new Options();

        // If an exception is thrown here, the exit code will not be 1, it will be 0.
        // while the exception handler is called.
        return new Options(value);
    }
}
```

And

```csharp
public void Handler(Options options)
{
    // Here many exceptions can occur, and most are handled 
    // according to the design; but the following exception,
    // in particular, will lead to application exit, after `UseExceptionHandler` 
    //is called but without returning `Invoke`. In other words, 
    //`ExitCode` will not be set, and it will be zero which is the default.
    throw new NotImplementedException();
}
```

**Comment by jbuhacoff** [comment]
I found this thread while searching for a way to let my command indicate the exit code and customize what happens when an exception is thrown. I was able to make it work after reading the discussion here so I'll just share what I did in case it helps anyone else.

```CSharp
struct Options
{
    public FileInfo? ConfigFile;
    public string? TestMode;
    /* etc. */
}

static int Main(string args[]) {
    var rootCommand = new RootCommand("Program name");
    var configFileOption = new Option<FileInfo>(name: "--config");
    var testModeOption = new Option<string>(name: "--test");
    rootCommand.AddOption(configFileOption);
    rootCommand.AddOption(testModeOption);
    /* etc. */
    rootCommand.SetHandler((context) =>
    {
        Options options = new Options()
        {
            ConfigFile = context.ParseResult.GetValueForOption(configFileOption),
            TestMode= context.ParseResult.GetValueForOption(testModeOption),
            /* etc. */
        };
        context.ExitCode = CommandWrapper(options);
    });
    var parser = new CommandLineBuilder(rootCommand)
        .UseDefaults()
        /* customizations here */
        .Build();
    return parser.Invoke(args);
}

static int CommandWrapper(Options options)
{
    try
    {
        return MyCommand(options);
    } catch (Exception e)
    {
        Console.WriteLine(e.Message);
        /* log the stack trace somewhere else for developers to see later */
        return 255;
    }
}

static int MyCommand(Options options)
{
    /* do the thing */
    return 0;
}
```

**Comment by CEbbinghaus** [comment]
Because it confused the hell out of me since I use the latest preview version. The way this is done (as of main today) is to use the `EnableDefaultExceptionHandler` property on the `CliConfiguration`
```cs
CliConfiguration config = new(command)
{
    EnableDefaultExceptionHandler = false
};
```

This feels like a big step forward although at the cost of a bad default. I strongly believe that the library should not be catching the exception and instead just allow people to turn on this behavior. Having your application silently exit for no reason is the absolute worst experience when you are trying to figure out what could have gone wrong.
